server/vertx/src/main/scala/sttp/tapir/server/vertx/decoders/VertxInputDecoders.scala

```diff
@@ -0,0 +1,125 @@
+import sttp.tapir.server.vertx.VertxEndpointOptions
+import sttp.tapir.server.vertx.encoders.VertxOutputEncoders
+import sttp.tapir.server.vertx.handlers.tryEncodeError
+import sttp.tapir.server.{DecodeFailureContext, DecodeFailureHandling}
+import sttp.tapir.{DecodeResult, Endpoint, EndpointIO, RawBodyType}
+
+import scala.reflect.ClassTag
+import scala.util.Random
+
+object VertxInputDecoders {
+
+  /**
+   * Decodes the inputs, the body if needed, and if it succeeds invokes the logicHandler
+   * @param endpoint the endpoint definition
+   * @param rc the RoutingContext
+   * @param logicHandler callback to execute if the inputs are decoded properly
+   * @param endpointOptions endpoint options (execution context, etc.)
+   * @param ect the error class to convert to, if the user recovers error
+   * @tparam E Error parameter type, if it should be caught and handled by the user
+   */
+  private [vertx] def decodeBodyAndInputsThen[E](
+    endpoint: Endpoint[_, E, _, _],
+    rc: RoutingContext,
+    logicHandler: Params => Unit,
+  )(implicit endpointOptions: VertxEndpointOptions, ect: Option[ClassTag[E]]): Unit = {
+    decodeBodyAndInputs(endpoint, rc) match {
+      case values: DecodeInputsResult.Values =>
+        InputValues(endpoint.input, values) match {
+          case InputValuesResult.Value(params, _) =>
+            logicHandler(params)
+          case InputValuesResult.Failure(_, failure) =>
+            tryEncodeError(endpoint, rc, failure)
+        }
+      case DecodeInputsResult.Failure(input, failure) =>
+        val decodeFailureCtx = DecodeFailureContext(input, failure)
+        endpointOptions.decodeFailureHandler(decodeFailureCtx) match {
+          case DecodeFailureHandling.NoMatch =>
+            endpointOptions.logRequestHandling.decodeFailureNotHandled(endpoint, decodeFailureCtx)(endpointOptions.logger)
+            rc.response.setStatusCode(404).end()
+          case DecodeFailureHandling.RespondWithResponse(output, value) =>
+            endpointOptions.logRequestHandling.decodeFailureHandled(endpoint, decodeFailureCtx, value)(endpointOptions.logger)
+            VertxOutputEncoders.apply(output, value)(endpointOptions)(rc)
+        }
+
+    }
+  }
+
+  private def decodeBodyAndInputs(e: Endpoint[_, _, _, _], rc: RoutingContext)
+                                         (implicit serverOptions: VertxEndpointOptions): DecodeInputsResult =
+    decodeBody(DecodeInputs(e.input, new VertxDecodeInputsContext(rc)), rc)
+
+  private def decodeBody(result: DecodeInputsResult, rc: RoutingContext)
+                        (implicit serverOptions: VertxEndpointOptions): DecodeInputsResult = {
+    result match {
+      case values: DecodeInputsResult.Values =>
+        values.bodyInput match {
+          case None => values
+          case Some(bodyInput@EndpointIO.Body(bodyType, codec, _)) =>
+            codec.decode(extractRawBody(bodyType, rc)) match {
```

Comment:
and this would propagate here, so that you would have a `Future[DecodeInputsResult]` etc.

Reply:
Thanks!